File: 백준 문제 풀이 모음/python/class 4/1629.py
import sys

a, b, c = map(int, sys.stdin.readline().split())

def mul(a,b) :
    if b == 1 :
        return a % c
    else :
        tmp = mul(a, b//2)
        if b % 2 == 0 :
            return (tmp * tmp) %c
        else :
            return (tmp * tmp * a) % c

print(mul(a,b))

File: 백준 문제 풀이 모음/kotlin/class 4/1629.kt


fun main() = with(System.`in`.bufferedReader()){
    val (a, b, c) = readLine().split(' ').toList().map { it.toLong() }

    val x = pow(a, b, c)
    print(x)
}

fun pow(base: Long, expo: Long, c :Long) : Long {
    if (expo == 1L) return base % c

    else {
        val x = pow(base, expo / 2, c)
        if (expo % 2 == 0L) {
            return (x * x) % c
        } else {
            return (((x * x) % c) * base)% c
        }
    }

}
